### ISSUE TITLE
Error while fine tune the googlenet model

### ISSUE THREAD

**Issue by noneface** [vlfeat/matconvnet#451]
Hi,
I am trying to use my own dataset to fine tune the googlenet model.
i run the cnn_train_dag.m function without changing the model's last two layers, however, it has some errors in cnn_train_dag.m as follows.

error  dagnn.DagNN/eval (line 80)
No variable of name 'input' could be found in the DAG.

and then i read the eval.m ,it seems that something wrong about getVarIndex(inputs(1:2:end)) in line 77.
About the inputs,it is:
   im = imdb.images.data(:,:,:,batch) ;
   labels = imdb.images.labels(1,batch) ;
   inputs = {'input', im, 'label', labels} ;
and in getVarIndex() function, it always return  NaN with my inputs.
I've also tried to use the imdb.mat file in examples/mnist, it also meets the same error, so I think my imdb.mat file has no problem. 

So, is there something wrong in getVarIndex() or missing some steps in my program?
And, according to https://github.com/vlfeat/matconvnet/issues/218, it is said that we should change the last two layers before fine tuning,  then how to change the last two layers in googlenet model?

**Comment by daofeng2007** [comment]
**## I'm also trying to finetune the googlenet using my own data, which have 7 classes. The following code can solve your problem, which I also encountered in the beginning.**

`modelPath = 'data/models/imagenet-googlenet-dag.mat' ;

net = dagnn.DagNN.loadobj(load(modelPath)) ;

%net.meta.inputSize = [net.meta.normalization.imageSize, 32] ;
net.meta.inputSize = [224,224,3,32] ;

net.meta.normalization.cropSize = net.meta.normalization.imageSize(1) / 256 ;
net.meta.normalization.averageImage = rgbMean;
net.meta.classes.name = {'class1','class2','class3','class4','class5','class6','class7'};
net.meta.classes.description = {'class1','class2','class3','class4','class5','class6','class7'};
net.meta.augmentation.jitterLocation = true ;
net.meta.augmentation.jitterFlip = true ;
net.meta.augmentation.jitterBrightness = double(0.1 * rgbDeviation) ;
net.meta.augmentation.jitterAspect = [2/3, 3/2] ;
if ~opts.batchNormalization
  lr = logspace(-2, -4, 60) ;
else
  lr = logspace(-1, -4, 20) ;
end

net.meta.trainOpts.learningRate = lr ;
net.meta.trainOpts.numEpochs = numel(lr) ;
net.meta.trainOpts.batchSize = 100 ;
net.meta.trainOpts.weightDecay = 0.0005 ;

net.renameVar(net.layers(1).inputs{1}, 'input')   **%this will solve you 'input' problem**
net.layers(153)=[]; %remove the softmax layer


% add following loss layers
net.addLayer('loss', dagnn.Loss('loss', 'classerror'), ...
             {'cls3_fc','label'}, 'objective') ;
net.addLayer('top1err', dagnn.Loss('loss', 'classerror'), ...
             {'cls3_fc','label'}, 'top1err') ;
net.addLayer('top5err', dagnn.Loss('loss', 'topkerror', ...
                                   'opts', {'topK',5}), ...
             {'cls3_fc','label'}, 'top5err') ;`


## ****However, I still got the following error****

train: epoch 01:   1/303:Error using vl_taccummex
A and B do not have the same number
of elements.

Error in vl_taccum (line 19)
  a = vl_taccummex(alpha, a, beta,
  b) ;

Error in
cnn_train_dag>accumulateGradients
(line 300)
        parDer =
        vl_taccum(1/batchSize,
        parDer, ...

Error in cnn_train_dag>processEpoch
(line 227)
    state = accumulateGradients(net,
    state, params, batchSize,
    parserv) ;

Error in cnn_train_dag (line 87)
    [net, state] = processEpoch(net,
    state, params, 'train') ;

Error in cnn_imagenet_test (line
140)
[net, info] = trainFn(net, imdb,
getBatchFn(opts, net.meta), ...

**Comment by FrancisYizhang** [comment]
@daofeng2007 
Did you solve the problem?
help~~~~~~~

**Comment by micklexqg** [comment]
same problem!

**Comment by maxiaoyuelun** [comment]
anybody solve this problem?

**Comment by firatozgenel** [comment]
@daofeng2007 
Hi, I have also had problems regarding with "Error using vl_taccummex". 
The problem is caused because of GoogleNet structure. It has three classifiers and three loss functions must be written to enable the calculation of derivatives for backpropagation. 

For training GoogleNet, I removed softmax layer and added the following codes for three loss layers. 

net.addLayer('loss1', ...
             dagnn.Loss('loss', 'softmaxlog') ,...
             {'cls1_fc2', 'label'}, ...
             'objective1') ;
         
net.addLayer('loss2', ...
             dagnn.Loss('loss', 'softmaxlog') ,...
             {'cls2_fc2', 'label'}, ...
             'objective2') ;
         
net.addLayer('loss3', ...
             dagnn.Loss('loss', 'softmaxlog') ,...
             {'cls3_fc', 'label'}, ...
             'objective3') ;

You will also need to set opts.DerOutputs as:

opts.derOutputs ={'objective1', 0.3,'objective2',0.3,'objective3',1} ;

The weight of the objective functions are given in the GoogleNet paper (https://arxiv.org/pdf/1409.4842.pdf)

In addition, I think you need to modify the output sizes of fully connected layers to match the number of classes. 

Hope this helps

**Comment by dengyuhk** [comment]
@firatozgenel , thanks for your answers, with your solution, I solved my problem.
